cmd/heimdalld/cmd/ve_decode.go

```diff
@@ -0,0 +1,431 @@
+	}
+
+	if len(br.Result.Block.Data.Txs) == 0 {
+		return nil, fmt.Errorf("no vote extensions found in the block")
+	}
+
+	veB64Str := br.Result.Block.Data.Txs[0]
+	veBytes, err := base64.StdEncoding.DecodeString(veB64Str)
+	if err != nil {
+		return nil, err
+	}
+
+	var voteExt abci.ExtendedCommitInfo
+	if err := goproto.Unmarshal(veBytes, &voteExt); err != nil {
+		return nil, err
+	}
+	return &voteExt, nil
+}
+
+func GetVEsFromBlockStore(height int64) (*abci.ExtendedCommitInfo, error) {
+	homeDir := viper.GetString(flags.FlagHome)
+	if homeDir == "" {
+		return nil, fmt.Errorf("home directory not set")
+	}
+
+	cometbftdb, err := cometbftDB.NewGoLevelDB("blockstore", path.Join(homeDir, "data"))
+	if err != nil {
+		return nil, err
+	}
+	blockStore := store.NewBlockStore(cometbftdb)
+	block := blockStore.LoadBlock(height)
+	if block == nil {
+		return nil, fmt.Errorf("block at height %d not found", height)
+	}
+
+	ves := block.Data.Txs[0] //nolint:staticcheck
+	if ves == nil {
+		return nil, fmt.Errorf("no vote extensions found in the block")
+	}
+
+	var voteExt abci.ExtendedCommitInfo
+	if err := voteExt.Unmarshal(ves); err != nil {
+		return nil, err
+	}
+
+	return &voteExt, nil
+}
+
+// BuildCommitJSON builds a JSON representation for the given ExtendedCommitInfo and block height.
+func BuildCommitJSON(height int64, ext *abci.ExtendedCommitInfo) ([]byte, error) {
+	data := CommitData{
+		Height: height,
+		Round:  ext.Round,
+		Votes:  make([]VoteData, len(ext.Votes)),
+	}
+
+	for i, v := range ext.Votes {
+		// Unmarshal sideTx extension
+		var stxs sidetxs.VoteExtension
+		if err := goproto.Unmarshal(v.VoteExtension, &stxs); err != nil {
+			return nil, err
+		}
+
+		vote := VoteData{
+			ValidatorAddr: "0x" + hex.EncodeToString(v.Validator.Address),
+			Power:         v.Validator.Power,
+			ExtSignature:  "0x" + hex.EncodeToString(v.ExtensionSignature),
+			BlockIDFlag:   v.BlockIdFlag.String(),
+		}
+
+		// SideTxResponses
+		for _, r := range stxs.SideTxResponses {
+			vote.SideTxs = append(vote.SideTxs, SideTxData{
+				TxHash: "0x" + hex.EncodeToString(r.TxHash),
+				Result: r.Result.String(),
+			})
+		}
+
+		// Milestone
+		if mp := stxs.MilestoneProposition; mp != nil {
+			hashes := make([]string, len(mp.BlockHashes))
+			for j, bh := range mp.BlockHashes {
+				hashes[j] = "0x" + hex.EncodeToString(bh)
+			}
+			vote.Milestone = &MilestoneData{
+				BlockHashes:      hashes,
+				StartBlockNumber: mp.StartBlockNumber,
+				ParentHash:       "0x" + hex.EncodeToString(mp.ParentHash),
+			}
+		}
+
+		// Non-RP vote extension: dummy vs checkpoint
+		if isDummy, _ := IsDummyNonRpVoteExtension(height, v.NonRpVoteExtension); isDummy {
+			vote.NonRpData = "0x" + hex.EncodeToString(v.NonRpVoteExtension)
+		} else {
+			msg, err := GetCheckpointMsg(v.NonRpVoteExtension)
+			if err != nil {
+				return nil, fmt.Errorf("error unpacking checkpoint: %w", err)
+			}
+			vote.NonRpData = CheckpointData{
+				Proposer:        msg.Proposer,
```

Comment:
You can use `FormatAddress` here (and everywhere you manipulate address, just for consistency)